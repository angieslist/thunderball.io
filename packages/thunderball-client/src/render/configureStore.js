import appReducers from '../reducers/index';
import { createLogger } from 'redux-logger';
import { applyMiddleware, createStore, compose, combineReducers } from 'redux';
import handleResetOnAction from './handleResetOnAction';

const isReactNative = typeof navigator === 'object' && navigator.product === 'ReactNative';
const defaultStoreOptions = {
  middleware: [],
  enhancers: [],
  reducers: appReducers,
  initialState: {},
  enableLogger: process.env.NODE_ENV !== 'production' && process.env.IS_BROWSER || isReactNative,
  resetOnAction: [],
};

export default function configureStore(options = {}, injectors = [], pageProps = {}) {
  // Injection point: beforeConfigureStore
  // Using the beforeConfigureStore, determine what the full storeOptions should be
  // by reducing each result of beforeConfigureStore with the previous
  const storeOptions = injectors
    .filter(injector => injector.beforeConfigureStore)
    .reduce((prevOptions, injector) => ({ ...prevOptions, ...injector.beforeConfigureStore(prevOptions, pageProps) }),
      // Create the base storeOptions from the defaults and those defined by the platform
      { ...defaultStoreOptions, ...options });

  // Logger must be the last middleware in chain.
  if (storeOptions.enableLogger) {
    storeOptions.middleware.push(createLogger({
      collapsed: true,
      // Convert objects to pure javascript
      stateTransformer: state => JSON.parse(JSON.stringify(state)),
    }));
  }

  // Add middleware as the first enhancer
  storeOptions.enhancers.unshift(applyMiddleware(...storeOptions.middleware));

  // Always push reduxDevTools to the end of the enhancer stack
  // This is only active in development mode and in clients with the devtools installed
  // Redux tools extension https://github.com/zalmoxisus/redux-devtools-extension
  if (process.env.NODE_ENV !== 'production' &&
    typeof window === 'object' && typeof window.devToolsExtension === 'function') {
    storeOptions.enhancers.push(window.devToolsExtension());
  }

  // Combine reducers and create store
  const combinedReducer = combineReducers(storeOptions.reducers);
  const store = createStore(
    handleResetOnAction(combinedReducer, storeOptions.initialState, storeOptions.resetOnAction),
    storeOptions.initialState,
    compose(...storeOptions.enhancers),
  );

  // Injection point: afterConfigureStore
  injectors
    .filter(injector => injector.afterConfigureStore)
    .forEach(injector => injector.afterConfigureStore(store, pageProps, storeOptions));

  return store;
}
