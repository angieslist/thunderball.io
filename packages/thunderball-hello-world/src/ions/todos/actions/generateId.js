export default function generateId() {
  return Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
}
