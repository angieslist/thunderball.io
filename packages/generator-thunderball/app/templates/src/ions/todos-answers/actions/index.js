import generateId from './generateId';
import { fetchTodos } from '../services/todosApi';

export const addTodo = (text, id = generateId(), completed = false) => ({
  type: 'ADD_TODO',
  id,
  text,
  completed,
});

export const setVisibilityFilter = filter => ({
  type: 'SET_VISIBILITY_FILTER',
  filter,
});

export const toggleTodo = id => ({
  type: 'TOGGLE_TODO',
  id,
});

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE',
};

export function fetchAndAddTodos() {
  return async (dispatch) => {
    try {
      const todos = await fetchTodos();
      todos.forEach((x) => {
        dispatch(addTodo(x.title, x.id, x.completed));
      });
    } catch (e) {
      // do nothing
    }
  };
}
