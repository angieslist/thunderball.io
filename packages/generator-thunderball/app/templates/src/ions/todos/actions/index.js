import generateId from './generateId';

export const addTodo = (text, id = generateId(), completed = false) => ({
  type: 'ADD_TODO',
  id,
  text,
  completed,
});

export const setVisibilityFilter = filter => ({
  type: 'SET_VISIBILITY_FILTER',
  filter,
});

export const toggleTodo = id => ({
  type: 'TOGGLE_TODO',
  id,
});

export const VisibilityFilters = {
  SHOW_ALL: 'SHOW_ALL',
  SHOW_COMPLETED: 'SHOW_COMPLETED',
  SHOW_ACTIVE: 'SHOW_ACTIVE',
};
