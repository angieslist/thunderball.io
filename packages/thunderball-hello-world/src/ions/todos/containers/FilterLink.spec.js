import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';

import FilterLink from './FilterLink';
import { setVisibilityFilter } from '../actions';

const setup = () => {
  const store = configureStore()({});
  const wrapper = shallow(
    <FilterLink filter="SHOW_ALL" store={store} >
      Test Link
    </FilterLink>,
  );

  return {
    store,
    wrapper,
  };
};

describe('FilterLink', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  test('sets the correct filter when clicked', () => {
    const { store, wrapper } = setup();
    expect(store.getActions()).toEqual([]);
    wrapper.find('Link').simulate('click');
    expect(store.getActions()).toEqual([setVisibilityFilter('SHOW_ALL')]);
  });
});
