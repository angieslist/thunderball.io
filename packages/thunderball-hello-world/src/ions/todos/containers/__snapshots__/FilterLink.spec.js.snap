// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterLink renders without crashing 1`] = `
<Link
  active={false}
  filter="SHOW_ALL"
  onClick={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
>
  Test Link
</Link>
`;
